Implement equals function for all types

thomcc opened the pull request:
Strict equality comparison, e.g. ===. So this might do the wrong thing for NaNs, depending on your perspective.  

I considered adding a matching approxEquals as well, but considering that I didn't need it, and that the way to do accurately has some overhead that a lot of people might find undesirable (the obvious `Math.abs(x-y) < epsilon` fails for large x and y, so you need to do  `Math.abs(x-y) < epsilon*Math.max(1.0, Math.abs(x), Math.abs(y))`)... I punted on it.

toji commented:
Sorry about the conflicts, probably wouldn't have happened if I had merged in order. This is good functionality, but I'm trying to decide if it will be more confusing than helpful. In my experience vast swaths of developers don't grok the imprecision inherent in floating point math, and would expect this to work on the results of some computation. (Like "if a * b * c === identity").

It's hard to argue that a feature should be kept out simply because it might be misunderstood, but maybe the name could be changed to better indicate it's function? exactlyEquals? isIdentical? reallyTruelyEqualsThisIsProbablyNotTheFunctionYouAreLookingFor? ;)

thomcc commented:
I could rename this to exactEquals, and add an equals() that works with an epsilon if that would help. Only question is where to get epsilon. Obvious choices are glMatrix.EPSILON or an argument. Probably the former?

toji commented:
exactEquals sounds fine, and for consistency we should use glMatrix.EPSILON. I think that will handle 99% of people's needs.

thomcc commented:
Added both. Also added an approximate equality comparison (named `equals` for consistency) to glMatrix. 

The comparison method is different from the one you use in the testing code (which is an `abs(a-b) < epsilon` check), but it helps out a great deal for numeric robustness. You can read about this [here](http://realtimecollisiondetection.net/pubs/Tolerances/). 

But, it might be unintuitive, and has a small performance cost associated with it, so I'd be willing to replace it with an `abs(a-b) < epsilon` check if you think that would be better.

toji commented:
Looks good. I'm not super concerned about the performance here, as this is the type of function that's more likely to be used during testing or other non-perf sensitive code. If we start hearing complaints it's worth re-evaluating but I don't see that happening. Thanks for your continued contributions!
